README.markdown
# GTI Tool - Lernwerkzeug für die Grundlagen der theoretischen Informatik

GTI Tool wurde als interaktives Lernwerkzeug zur Vorlesung Grundlagen der theoretischen Informatik entwickelt, das den Studierenden helfen soll, den Vorlesungs- und Übungsinhalt interaktiv zu erlernen, und damit eine solide Grundlage für das Verständnis der vermittelten Konzepte zu erwerben. Darüberhinaus soll es die Prüfungsvorbereitung erleichtern, da Studierende nun leicht selbst überprüfen können, ob sie ein bestimmtes Konzept verstanden haben, statt sich auf die in der Vorlesung und Übung besprochenen Beispiele als Referenz beschränken zu müssen.


gtitool.jgraph/ChangeLog
11-SEP-2007: JGraph 5.10.1.5

- DefaultGraphModel update methods call superclass methods only
- Fixed mouse event offset while dragging edge labels
- addPort() method changed to return newly created port
- Adds parallel router to core

30-JUL-2007: JGraph 5.10.1.4

- Fixes snapping to grid for moving more than MAXCELLS cells
- Fixes rounding errors in BasicGraphUI.RootHandle.mouseDragged
- Fixes relative edge label positions
- Adds handling of window close event in EditorGraph example

16-JUL-2007: JGraph 5.10.1.3

- Fixes background repainting on XOR dragging issues

30-MAY-2007: JGraph 5.10.1.2

- Fixes out of visible area painting issues

24-MAY-2007: JGraph 5.10.1.1

- Fixes offscreen positioning for relatively positioned container
- Corrects user manual font issue.

07-MAY-2007: JGraph 5.10.1.0

- Fixes marquee selection artifacts on autoscrolling
- Uses absolute offsets for better combined edge label positions
- Adds EdgeHandle.getRelativeLabelPosition for relative label positioning
- Adds EdgeRenderer.getRelativeLabelPosition for better label positions
- Added GraphLayoutCache parameter to refresh, update and route methods
- Adds getPortSize() and setPortSize() so custom ports may override static value
- Adds check for edge visibility in getNeighbours
- Ungrouping now only removes ports of removed cells
      see bug #36 - http://www.jgraph.com/tracker/bug.php?op=show&bugid=36
- Added groupsEditable flag to JGraph
- Fixed scale centering for zoom less than 1
- DefaultRouter does no longer add control points that overlap source or target
- Fixes ordering issues in model
- Adds removeEmptyGroups switch in DefaultGraphModel
- Adds moveBeyondGraphBounds to JGraph to constrain dragging to graph bounds
- Adds constrainDrag() in rootHandle for controlling where dragging is allowed
- build.xml references examples main in com rather than org package
- isVertex now returns false for null vertex parameter
      see bug #28 - http://www.jgraph.com/tracker/bug.php?op=show&bugid=28
- Removed setting of initialLocation in RootHandle.mouseReleased to null
      see bug #27 - http://www.jgraph.com/tracker/bug.php?op=show&bugid=27
- Fixes edge label repaint on Macs
- Fixes marquee color and repaint on Macs
- Selection on insertion defaults to false
- Adds link attribute to GraphConstants

03-AUG-2006: JGraph-5.10.0.1 Beta 2

- Fixes possible NPE during grid repainting
- Fixes non-accelerated buffered image not always created when graph changes size
- Corrects background repaint to allow for clip bounds
- Reduces repaint area using clip bounds
- Corrects clip bounds bug when renderering non-accelerated double buffer

30-JUL-2006: JGraph-5.10.0.0 Beta 1

- Changes selection in default selection model from ArrayList to LinkedHashSet
- Improves double buffering image retention across operations
- Minimum Java version now 1.4

25-JUL-2006: JGraph-5.9.2.0-Cardiff

- Corrects old value on selection enable change in JGraph
- Improves performance when removing cells
- Improves performance when making cells invisible
- Adds static helper method DefaultGraphModel.getRootsAsCollection()
- Adds support for Component to be used as graph background

11-JUL-2006: JGraph-5.9.1.0-Berlin

- Adds background image support
- Adds port.gif to free distribution
- Fixes incorrect first paint in GraphEdX

06-JUL-2006: JGraph-5.9.0.0-Berlin

- Moves double buffering image to JGraph to share more effectively
- Removes labels caching and label bounds methods from EdgeView

22-JUN-2006: JGraph-5.8.3.2-Munich

- Changes getInsertionOffset to protected

21-JUN-2006: JGraph-5.8.3.1-Munich

- Adds case in handleAttributes to allow for correct removal of the user object
- Adds getInsertionOffset hook in GraphTransferHandler to position cloned cells
- Corrects NPE possibly caused in EdgeRenderer.getLabelPosition()
- Adds JPanel vertex example

14-JUN-2006: JGraph-5.8.3.0-Munich

- Various example additions and improvements
- Fixes leak on Macs by not using XOR painting in marquee, handles & port renderer
- Corrected memory display on GraphEd status bar

07-JUN-2006: JGraph-5.8.2.3-Malibu

- Add Port Labels example

29-MAY-2006: JGraph-5.8.2.2-Los Angeles

- Fixes infinite loop when null cell passed into remove()
- Deprecates GraphConstants.getValue, use GraphModel instead

23-MAY-2006: JGraph-5.8.2.1-San Francisco

- Adds manual dispose() calls to assist Apple JVM GC reclaim memory

17-MAY-2006: JGraph-5.8.2-San Francisco

- Adds com.jgraph examples to free distribution
- Adds getPartial in GraphLayoutCache to fix persistence problem

02-MAY-2006: JGraph-5.8.1.1-Vienna

- Fixes preview of children for context edges in GraphContext

30-APR-2006: JGraph-5.8.1.0-Vienna

- Fixes XMLEncoding/Serialization for file I/O in GraphEdX
- Adds serialization for weak hash map in GraphLayoutCache
- Scaling now centers zoom correctly
- Adds JGraph.getCenterPoint method
- EdgeView.getPointCount checks for points being null before using
- Calls marquee handler for toggle events if isForceMarquee returns true
- Adds edgeLabelMovable switch to graph to prevent edge label movement
- Corrects NPE editing null edge labels when setAutoSizeOnValueChange is true
- Makes AbstractCellView.editor public so it can be changed

16-MAR-2006: JGraph-5.8.0.0-Vienna

- Uses weak graph reference in EdgeRenderer
- Adds MyCellView demo demonstrating cell view factory usage
- Adds GraphSelectionDemo, demonstrating graph selection events

28-FEB-2006: JGraph-5.7.4.8-London

- Adds PREVIEW_EDGE_GROUP global switch to GraphContext class
- Uses CompoundVertexView for groups in GraphEd (Thanks to Jerry Pulley!)
- Creating default bounds in GraphEd.group for new group cells
- Changes AbstractCellView.translate and scale to public visibility
- Disposes graphics in JGraph.getImage to clean-up resources

20-FEB-2006: JGraph-5.7.4.7-London

- User Manual freely available

30-JAN-2006: JGraph-5.7.4.6-London

- Fixes minor problems with port magic

jflex/source/JFlex/sym.java
// ----------------------------------------------------
// The following code was generated by CUP v0.10k
// Sun Nov 07 14:07:52 GMT+10:00 2004
// ----------------------------------------------------

package JFlex;


/** CUP generated interface containing symbol constants. */
@SuppressWarnings (
{ "all" } )
public interface sym
{

  /* terminals */
  public static final int DELIMITER = 9;


  public static final int EOFRULE = 25;


  public static final int STRING_I = 45;


  public static final int CHAR = 39;


  public static final int CONCAT = 44;


  public static final int MACROUSE = 41;


  public static final int LOWERCLASS = 24;


  public static final int EQUALS = 10;


  public static final int CCLASSNOT = 43;


  public static final int REPEAT = 31;


  public static final int TILDE = 38;


  public static final int CHAR_I = 46;


  public static final int STAR = 32;


  public static final int BAR = 34;


  public static final int FULL = 16;


  public static final int LOOKAHEAD = 27;


  public static final int LESSTHAN = 12;


  public static final int BANG = 37;


  public static final int COMMA = 11;


  public static final int JLETTERDIGITCLASS = 20;


  public static final int PLUS = 33;


  public static final int USERCODE = 30;


  public static final int EOF = 0;


  public static final int error = 1;


  public static final int DASH = 8;


  public static final int CLOSEBRACKET = 3;


  public static final int HAT = 4;


  public static final int JLETTERCLASS = 19;


  public static final int IDENT = 29;


  public static final int ACTION = 28;


  public static final int CLOSECLASS = 7;


  public static final int DIGITCLASS = 22;


  public static final int LBRACE = 14;


  public static final int REGEXPEND = 18;


  public static final int OPENCLASS = 6;


  public static final int OPENBRACKET = 2;


  public static final int POINT = 36;


  public static final int LETTERCLASS = 21;


  public static final int MORETHAN = 13;


  public static final int DOLLAR = 5;


  public static final int RBRACE = 15;


  public static final int STRING = 40;


  public static final int QUESTION = 35;


  public static final int UPPERCLASS = 23;


  public static final int NOACTION = 26;


  public static final int CCLASS = 42;


  public static final int UNICODE = 17;
}
